x/evm/keeper/migrate.go

```diff
@@ -3,68 +3,115 @@ package keeper
+// Migrate8to9 returns the handler that performs in-place store migrations
+func Migrate8to9(bk *BaseKeeper, n types.Nexus) func(ctx sdk.Context) error {
 	return func(ctx sdk.Context) error {
-		chains := slices.Filter(n.GetChains(ctx), func(chain exported.Chain) bool { return chain.Module == types.ModuleName })
-		for _, chain := range chains {
-			if err := migrateBurnerInfoForChain(ctx, k, chain, burnerAddrPrefixDeprecated); err != nil {
-				return err
-			}
-			if err := migrateBurnerInfoForChain(ctx, k, chain, strings.ToLower(burnerAddrPrefixDeprecated)); err != nil {
-				return err
-			}
-
-			k.Logger(ctx).Info(fmt.Sprintf("migrated all burner info keys for chain %s", chain.Name))
+		for _, chain := range slices.Filter(n.GetChains(ctx), types.IsEVMChain) {
+			ck := funcs.Must(bk.ForChain(ctx, chain.Name)).(chainKeeper)
 
+			migrateDeposits(ctx, ck, types.DepositStatus_Confirmed)
+			migrateDeposits(ctx, ck, types.DepositStatus_Burned)
 		}
-		k.Logger(ctx).Info("burner info keys migration complete")
+
 		return nil
 	}
 }
 
-func migrateBurnerInfoForChain(ctx sdk.Context, k *BaseKeeper, chain exported.Chain, oldKey string) error {
-	ck, err := k.forChain(ctx, chain.Name)
-	if err != nil {
-		return err
+func migrateDeposits(ctx sdk.Context, ck chainKeeper, status types.DepositStatus) {
+	var iteratedDepositCount, ignoredDepositCount uint64
+
+	var prefix key.Key
+	switch status {
+	case types.DepositStatus_Confirmed:
+		prefix = key.FromStr(confirmedDepositPrefixDeprecated)
+	case types.DepositStatus_Burned:
+		prefix = key.FromStr(burnedDepositPrefixDeprecated)
 	}
 
-	// migrate in batches so memory pressure doesn't become too large
-	keysToDelete := make([][]byte, 0, 1000)
-	for {
-		iterBurnerAddr := ck.getStore(ctx).Iterator(utils.KeyFromStr(oldKey))
+	iter := ck.getStore(ctx).IteratorNew(prefix)
+	defer utils.CloseLogError(iter, ck.Logger(ctx))
 
-		if !iterBurnerAddr.Valid() {
-			return iterBurnerAddr.Close()
+	for ; iter.Valid(); iter.Next() {
+		iteratedDepositCount++
+
+		var deposit types.ERC20Deposit
+		iter.UnmarshalValue(&deposit)
+
+		transferEvents := getTransferEventsByTxIDAndAddress(ctx, ck, deposit.TxID, deposit.BurnerAddress)
+		if len(transferEvents) == 0 {
+			// Deposits from the time when we had not started doing event processing.
+			// Their log indexes cannot be retrieved anymore and are therefore ignored.
+			ignoredDepositCount++
+			continue
 		}
-		var burnerInfo types.BurnerInfo
-		for ; iterBurnerAddr.Valid() && len(keysToDelete) < 1000; iterBurnerAddr.Next() {
-			iterBurnerAddr.UnmarshalValue(&burnerInfo)
-			ck.SetBurnerInfo(ctx, burnerInfo)
-			keysToDelete = append(keysToDelete, iterBurnerAddr.Key())
+
+		rawKey := iter.Key()
+		defer ck.getStore(ctx).DeleteRaw(rawKey)
+
+		for _, event := range transferEvents {
+			_, existingDepositStatus, ok := ck.GetDeposit(ctx, event.TxID, event.Index)
+			if ok && existingDepositStatus == status {
+				continue
+			}
```

Comment:
when is this case going to happen? Is this not exclusive with the getByIDAndAddress call?

Reply:
based on my understanding, there could be two deposits that had the same tx ID but different log index that were both burned under the old prefix. we'll iterate over both in the main loop. the second one would cause the same transfer events to be retrieved again (since that's done by the tx ID and burner address) and trigger this no-op check. We can't filter by the Log Index when we get transfer events, because the old deposits have it default to `0`.

Reply:
ah makes sense

Reply:
yup Milap got it right.